File: solutions/0125-valid-palindrome/solution.py
class Solution:
    def isPalindrome(self, s):
        # s= "".join(char.lower() for char in s if char.isalnum())
        # left=0
        # right= len(s)-1
        # while left< right:
        #     if s[left]!=s[right]:
        #         return False
        #     else:
        #         left+=1
        #         right-=1
        # return True

        left=0
        right= len(s)-1
        while left< right:
            if not s[left].isalnum():
                left+=1
                continue
            if not s[right].isalnum():
                right-=1
                continue
            if s[left].lower() != s[right].lower():
                return False
            else:
                left+=1
                right-=1
        return True

File: solutions/0125-valid-palindrome/solution.js
var isPalindrome = function(s) {
    let left = 0;
    let right = s.length - 1;

    while (left < right) {
        // skip non-alphanumeric on left
        while (left < right && !isAlphaNum(s[left])) {
            left++;
        }

        // skip non-alphanumeric on right
        while (left < right && !isAlphaNum(s[right])) {
            right--;
        }

        // compare lowercase characters
        if (s[left].toLowerCase() !== s[right].toLowerCase()) {
            return false;
        }

        left++;
        right--;
    }

    return true;
};

// helper function: checks if char is alphanumeric
function isAlphaNum(c) {
    return /[a-z0-9]/i.test(c);
}

    
    // O(n) & O(n)
    // let clean = s.toLowerCase().replace(/[^a-z0-9]/g, "");
    // let reversed = clean.split("").reverse().join("");
    // return clean === reversed;
    

